pkg/execution/state/redis_state/queue_processor.go

```diff
@@ -716,20 +713,39 @@ func (q *queue) processPartition(ctx context.Context, p *QueuePartition, shard *
 	// within 5ms of each other, we fetch them in order but we may process them out of
 	// order, depending on how long it takes for the item to pass through the channel
 	// to the worker, how long Redis takes to lease the item, etc.
-	fetch := time.Now().Truncate(time.Second).Add(PartitionLookahead)
+	fetch := getNow().Truncate(time.Second).Add(PartitionLookahead)
 	queue, err := q.Peek(peekCtx, p.Queue(), fetch, q.peekSize())
 	if err != nil {
 		return err
 	}
 
-	var processErr error
-	var concurrencyLimitReached bool
+	var (
+		processErr error
+
+		// These flags are used to handle partition rqeueueing.
+		ctrSuccess     int32
+		ctrConcurrency int32
+		ctrRateLimit   int32
+	)
 
 	// Record the number of partitions we're leasing.
 	q.int64counter(ctx, "inngest_queue_partition_lease_total", 1)
 
+	// staticTime is used as the processing time for all items in the queue.
+	// We process queue items sequentially, and time progresses linearly as each
+	// queue item is processed.  We want to use a static time to prevent out-of-order
+	// processing with regards to things like rate limiting;  if we use time.Now(),
+	// queue items later in the array may be processed before queue items earlier in
+	// the array depending on eg. a rate limit becoming available half way through
+	// iteration.
+	staticTime := getNow()
+
 ProcessLoop:
 	for _, qi := range queue {
+		// TODO: Create an in-memory mapping of rate limit keys that have been hit,
+		//       and don't bother to process if the queue item has a limited key.  This
+		//       lessens work done in the queue, as we can `continue` immediately.
```

Comment:
why not store the map in redis instead with a TTL for those keys?
don't remember if you can expire items in a HSET though.

if you can, that can be shared across all executors so each don't have to recreate the map themselves.

Reply:
That's not really any different from checking a rate limit, which is a single get and set :)

This would prevent us from accessing the backend overall, reducing _any_ round trips as a small optimization in this core loop.